lxd/storage_volumes.go

```diff
@@ -0,0 +1,530 @@
+	// Get the name of the storage volume.
+	volumeName := mux.Vars(r)["name"]
+	if volumeName == "" {
+		return BadRequest(fmt.Errorf("You must provide a name for the volume."))
+	}
+
+	// Get the name of the storage pool the volume is supposed to be
+	// attached to.
+	poolName := mux.Vars(r)["pool"]
+	if poolName == "" {
+		return BadRequest(fmt.Errorf("You must provide the name of the pool the volume belongs to."))
+	}
+
+	// Get the name of the volume type.
+	volumeTypeName := mux.Vars(r)["type"]
+	if volumeTypeName == "" {
+		return BadRequest(fmt.Errorf("You must provide a storage volume type of the storage volume."))
+	}
+	// Convert the volume type name to our internal integer representation.
+	volumeType, err := storagePoolVolumeTypeNameToType(volumeTypeName)
+	if err != nil {
+		return BadRequest(err)
+	}
+	// Check that the storage volume type is valid.
+	if !shared.IntInSlice(volumeType, supportedVolumeTypes) {
+		return BadRequest(fmt.Errorf("Invalid storage volume type %s.", volumeTypeName))
+	}
+
+	// Get the ID of the storage pool the storage volume is supposed to be
+	// attached to.
+	poolID, pool, err := dbStoragePoolGet(d.db, poolName)
+	if err != nil {
+		if err == NoSuchObjectError {
+			return BadRequest(fmt.Errorf("The storage pool does not exist."))
+		}
+		return InternalError(err)
+	}
+
+	// Get the existing storage volume.
+	_, volume, err := dbStoragePoolVolumeGetType(d.db, volumeName, volumeType, poolID)
+	if err != nil {
+		if err == NoSuchObjectError {
+			return BadRequest(fmt.Errorf("The storage volume does not exist."))
+		}
+		return InternalError(err)
+	}
+
+	// Validate the ETag
+	etag := []interface{}{volume.VolumeName, volume.VolumeType, volume.VolumeUsedBy, volume.VolumeConfig}
+
+	err = etagCheck(r, etag)
+	if err != nil {
+		return PreconditionFailed(err)
+	}
+
+	req := api.StorageVolume{}
+	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+		return BadRequest(err)
+	}
+
+	if req.VolumeConfig == nil {
+		req.VolumeConfig = map[string]string{}
+	}
+
+	for k, v := range volume.VolumeConfig {
+		_, ok := req.VolumeConfig[k]
+		if !ok {
+			req.VolumeConfig[k] = v
+		}
+	}
+
+	// Validate the configuration
+	err = storageVolumeValidateConfig(volumeName, req.VolumeConfig, pool)
+	if err != nil {
+		return BadRequest(err)
+	}
+
+	err = storagePoolVolumeUpdate(d, poolName, req.VolumeName, volumeType, req.VolumeConfig)
+	if err != nil {
+		return InternalError(err)
+	}
+
+	return EmptySyncResponse
+}
+
+// /1.0/storage-pools/{pool}/volumes/{type}/{name}
+func storagePoolVolumeTypeDelete(d *Daemon, r *http.Request) Response {
+	// Get the name of the storage volume.
+	volumeName := mux.Vars(r)["name"]
+	if volumeName == "" {
+		return BadRequest(fmt.Errorf("You must provide a name for the volume."))
+	}
+
+	// Get the name of the storage pool the volume is supposed to be
+	// attached to.
+	poolName := mux.Vars(r)["pool"]
+	if poolName == "" {
```

Comment:
I don't think you need to check for "" here. Relevant IRC discussion:

```
2017-02-13 10:19:29 tych0 brauner: one other comment, it looks like you check for "" on mux.Vars()
2017-02-13 10:19:52 tych0 brauner: i think that may have been coming from how we check for "" on req.FormValue()
2017-02-13 10:20:23 brauner yeah
2017-02-13 10:20:25 tych0 the difference here is that FormValue() gives you stuff that's part of the query string, which is optional. the mux.Vars() call will never give you "" unless you've typo'd the contents
2017-02-13 10:20:37 brauner ah, cool.
2017-02-13 10:20:48 tych0 because the request wouldn't get routed there without a value
2017-02-13 10:20:57 brauner sorry, i'm zoning out from time to time because i'm working on the fixes
2017-02-13 10:20:59 tych0 so i think you can drop all o fthose BadRequest() checks if you want
2017-02-13 10:21:01 brauner about the db stuff
2017-02-13 10:21:05 brauner i agree.
2017-02-13 10:21:09 tych0 brauner: np, i'm just brain dumping everything in here
2017-02-13 10:21:23 tych0 brauner: i can post a comment about the mux.Vars() thing if you want
2017-02-13 10:21:26 brauner however, i fear that this might be too late for the storage branch. (i'm not sure how much work that is)
2017-02-13 10:21:31 brauner tych0: please do
```